ext/oci8/oci8.stub.php

```diff
@@ -0,0 +1,734 @@
+<?php
+
+/**
+ * @param resource $statement_resource
+ * @param string $column_name
+ * @param mixed $variable
+ * @param int $type
+ * @return bool
+ */
+function oci_define_by_name($statement_resource, string $column_name, &$variable, int $type = 0): bool {}
+
+/**
+ * @param resource $statement_resource
+ * @param string $column_name
+ * @param mixed $variable
+ * @param int $maximum_length
+ * @param int $type
+ * @return bool
+ */
+function oci_bind_by_name($statement_resource, string $column_name, &$variable, int $maximum_length = -1, int $type = 0): bool {}
+
+/**
+ * @param resource $statement_resource
+ * @param string $column_name
+ * @param mixed $variable
+ * @param int $maximum_array_length
+ * @param int $maximum_item_length
+ * @param int $type
+ * @return bool
+ */
+function oci_bind_array_by_name($statement_resource, string $column_name, &$variable, int $maximum_array_length, int $maximum_item_length = -1, int $type = SQLT_AFC): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return bool
+ */
+function oci_free_descriptor(OCI_Lob $lob_descriptor): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param string $data
+ * @param int $offset
+ * @return bool
+ */
+function oci_lob_save(OCI_Lob $lob_descriptor, string $data, int $offset = 0): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param $filename @todo path?
+ * @return bool
+ */
+function oci_lob_import(OCI_Lob $lob_descriptor, $filename): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return string|false @todo return false
+ */
+function oci_lob_load(OCI_Lob $lob_descriptor) {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param int $length
+ * @return string|false @todo return false
+ */
+function oci_lob_read(OCI_Lob $lob_descriptor, int $length) {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return bool
+ */
+function oci_lob_eof(OCI_Lob $lob_descriptor): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return int|false @todo return false
+ */
+function oci_lob_tell(OCI_Lob $lob_descriptor) {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return bool
+ */
+function oci_lob_rewind(OCI_Lob $lob_descriptor): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param int $offset
+ * @param int $whence
+ * @return bool
+ */
+function oci_lob_seek(OCI_Lob $lob_descriptor, int $offset, int $whence = 0): bool {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return int|false @todo return false
+ */
+function oci_lob_size(OCI_Lob $lob_descriptor) {}
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param string $string
+ * @param int $length
+ * @return int|false @todo return false
+ */
+function oci_lob_write(OCI_Lob $lob_descriptor, string $string, int $length = 0) {}
```

Comment:
It seems to me that the default value of `$length` is not right here. Even though it's apparently `0`, we can't document this yet, since the behaviour changes depending on whether we provide the 3rd parameter due to the `if (ZEND_NUM_ARGS() == ...` checks). Ideally, the behaviour should depend on the value solely.

I think there's two possibility now: either we should mark the default value as `UNKNOWN` in the stub for now, or we should fix the default value handling. I'm not sure if we can use `0` as a default value because that seems to be a breaking change for me. @nikic please correct me if I'm wrong. Should we use `null` as default? Maybe `-1`?

Reply:
I'd go for null default here (or UNKNOWN for now). I looked at what fwrite() does, but unfortunately that one is also arg count dependent, so not a good reference :)

Reply:
@kocsismate @nikic I made $length nullable and made null the default value, is that ok?

Reply:
Currently, the default value is neither `0`, nor `null`, not anything else because default value handling is not right in the code :) In these situations we document `UNKNOWN`. In order to be eligible for documenting `null`, the implementation also has to be adapted. Here's an example PR which does the same: https://github.com/php/php-src/pull/5516/

This PR is the part of a crusade which is currently going on against `UNKNOWN` default values. Right now, we have around 280 of such default values, but it used to be even more, probably 400-ish.

So I think it's perfectly ok if you just use `UNKNOWN` for now. However, if you are up for the challenge, you can also fix these after the ZPP-related PR is merged.

Reply:
I removed nullability and made the default value `UNKNOWN`.